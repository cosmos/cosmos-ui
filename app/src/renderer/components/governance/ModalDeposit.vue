<template>
  <action-modal
    id="modal-deposit"
    ref="actionModal"
    title="Deposit"
    class="modal-deposit"
    @close-action-modal="close"
  >
    <tm-form-group
      :error="
        $v.amount.$error && $v.amount.$invalid && (amount > 0 || balance === 0)
      "
      class="action-modal-form-group"
      field-id="amount"
      field-label="Amount"
    >
      <span class="input-suffix">{{ denom }}</span>
      <tm-field v-focus id="amount" v-model="amount" type="number" />
      <tm-form-msg
        v-if="
          $v.amount.$error && !$v.amount.between && amount > 0 && balance > 0
        "
        :max="$v.amount.$params.between.max"
        :min="$v.amount.$params.between.min"
        name="Amount"
        type="between"
      />
      <tm-form-msg
        v-else-if="balance === 0"
        :msg="`doesn't hold any ${denom}s`"
        name="Wallet"
        type="custom"
      />
      <hr />
    </tm-form-group>
    <tm-form-group
      class="action-modal-form-group"
      field-id="password"
      field-label="Password"
    >
      <tm-field
        id="password"
        v-model="password"
        type="password"
        placeholder="Password"
      />
      <tm-form-msg
        v-if="$v.password.$error && !$v.password.required"
        name="Password"
        type="required"
      />
    </tm-form-group>
    <div class="action-modal-footer">
      <tm-btn
        v-if="sending"
        value="Sending..."
        disabled="disabled"
        color="primary"
      />
      <tm-btn
        v-else-if="!connected"
        value="Connecting..."
        disabled="disabled"
        color="primary"
      />
      <tm-btn
        v-else
        id="submit-deposit"
        color="primary"
        value="Submit Deposit"
        @click.native="validateForm"
      />
    </div>
  </action-modal>
</template>

<script>
import { mapGetters } from "vuex"
import ClickOutside from "vue-click-outside"
import { required, between } from "vuelidate/lib/validators"
import Modal from "common/TmModal"
import TmBtn from "common/TmBtn"
import TmField from "common/TmField"
import TmFormGroup from "common/TmFormGroup"
import TmFormMsg from "common/TmFormMsg"
import ActionModal from "common/ActionModal"

const isInteger = amount => Number.isInteger(amount)

export default {
  name: `modal-deposit`,
  components: {
    ActionModal,
    Modal,
    TmBtn,
    TmField,
    TmFormGroup,
    TmFormMsg
  },
  directives: {
    ClickOutside
  },
  props: {
    proposalId: {
      type: [Number, String],
      required: true
    },
    proposalTitle: {
      type: String,
      required: true
    },
    denom: {
      type: String,
      required: true
    }
  },
  data: () => ({
    amount: 0,
    password: ``,
    sending: false
  }),
  computed: {
    ...mapGetters([`wallet`, `connected`]),
    balance() {
      // TODO: refactor to get the selected coin when multicoin deposit is enabled
      if (!this.wallet.loading && !!this.wallet.balances.length) {
        let balance = this.wallet.balances.find(
          coin => coin.denom === this.denom
        )
        if (balance) return parseFloat(balance.amount)
      }
      return 0
    }
  },
  validations() {
    return {
      amount: {
        required,
        isInteger,
        between: between(1, this.balance)
      },
      password: {
        required
      }
    }
  },
  methods: {
    close() {
      this.$emit(`update:showModalDeposit`, false)
    },
    validateForm() {
      this.$v.$touch()

      if (!this.$v.$invalid) {
        this.submitForm()
      }
    },
    async submitForm() {
      this.sending = true

      await this.$refs.actionModal.submit(async () => {
        // TODO: support multiple coins
        await this.$store.dispatch(`submitDeposit`, {
          proposal_id: this.proposalId,
          amount: [
            {
              amount: String(this.amount),
              denom: this.denom
            }
          ],
          password: this.password
        })

        this.$store.commit(`notify`, {
          title: `Successful deposit!`,
          body: `You have successfully deposited your ${
            this.denom
          }s on proposal #${this.proposalId}`
        })
      }, `Depositing failed`)

      this.sending = false
    }
  }
}
</script>
