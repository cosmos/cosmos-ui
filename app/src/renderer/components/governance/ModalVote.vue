<template>
  <action-modal
    id="modal-vote"
    ref="actionModal"
    :submit-fn="submitForm"
    :simulate-fn="simulateForm"
    :validate="validateForm"
    title="Vote"
    class="modal-vote"
    submission-error-prefix="Voting failed"
    @close="clear"
  >
    <tm-form-group class="action-modal-group vote-options">
      <tm-btn
        id="vote-yes"
        :class="[vote === `Yes` ? 'active' : '']"
        :disabled="lastVoteOption === `Yes`"
        color="secondary"
        value="Yes"
        size="md"
        @click.native="vote = 'Yes'"
      />
      <tm-btn
        id="vote-no"
        :class="[vote === `No` ? 'active' : '']"
        :disabled="lastVoteOption === `No`"
        color="secondary"
        value="No"
        size="md"
        @click.native="vote = 'No'"
      />
      <tm-btn
        id="vote-veto"
        :class="[vote === `NoWithVeto` ? 'active' : '']"
        :disabled="lastVoteOption === `NoWithVeto`"
        color="secondary"
        value="No With Veto"
        size="md"
        @click.native="vote = 'NoWithVeto'"
      />
      <tm-btn
        id="vote-abstain"
        :class="[vote === `Abstain` ? 'active' : '']"
        :disabled="lastVoteOption === `Abstain`"
        color="secondary"
        value="Abstain"
        size="md"
        @click.native="vote = 'Abstain'"
      />
      <tm-form-msg
        v-if="$v.vote.$error && !$v.vote.required"
        name="Vote"
        type="required"
      />
    </tm-form-group>
  </action-modal>
</template>

<script>
import { mapGetters } from "vuex"
import { required } from "vuelidate/lib/validators"
import { uatoms } from "../../scripts/num.js"
import ActionModal from "common/ActionModal"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormMsg from "common/TmFormMsg"

const isValid = option =>
  option === `Yes` ||
  option === `No` ||
  option === `NoWithVeto` ||
  option === `Abstain`

export default {
  name: `modal-vote`,
  components: {
    ActionModal,
    TmBtn,
    TmFormGroup,
    TmFormMsg
  },
  props: {
    proposalId: {
      type: [Number, String],
      required: true
    },
    proposalTitle: {
      type: String,
      required: true
    },
    lastVoteOption: {
      default: undefined,
      type: String
    }
  },
  data: () => ({
    vote: null
  }),
  computed: {
    ...mapGetters([`bondDenom`])
  },
  validations() {
    return {
      vote: {
        required,
        isValid
      }
    }
  },
  methods: {
    open() {
      this.$refs.actionModal.open()
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()

      this.vote = null
    },
    async simulateForm() {
      return await this.$store.dispatch(`simulateVote`, {
        proposal_id: this.proposalId,
        option: this.vote
      })
    },
    async submitForm(gasEstimate, gasPrice, password, submitType) {
      await this.$store.dispatch(`submitVote`, {
        proposal_id: this.proposalId,
        option: this.vote,
        gas: String(gasEstimate),
        gas_prices: [
          {
            amount: String(uatoms(gasPrice)),
            denom: this.bondDenom
          }
        ],
        password,
        submitType
      })

      this.$store.commit(`notify`, {
        title: `Successful vote!`,
        body: `You have successfully voted ${this.vote} on proposal #${
          this.proposalId
        }`
      })
    }
  }
}
</script>

<style>
.proposal-title {
  font-weight: 500;
}

.action-modal-group.tm-form-group.vote-options {
  padding: 2rem 0;
}

.radio-container {
  display: flex;
  align-items: center;
  padding: 0.25rem 0;
}

.radio-container label {
  padding-left: 0.5rem;
}

.vote-options button {
  margin: 0;
  min-width: 49%;
}

.vote-options button span {
  margin: 0.25rem;
}

.vote-options button.active span {
  background: var(--tertiary);
}
</style>
