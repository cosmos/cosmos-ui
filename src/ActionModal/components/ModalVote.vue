<template>
  <ActionModal
    id="modal-vote"
    ref="actionModal"
    :validate="validateForm"
    title="Vote"
    class="modal-vote"
    submission-error-prefix="Voting failed"
    :transaction-data="transactionData"
    :notify-message="notifyMessage"
    @close="clear"
  >
    <TmFormGroup class="action-modal-group vote-options">
      <TmBtn
        id="vote-yes"
        :class="[vote === `Yes` ? 'active' : '']"
        :disabled="lastVoteOption === `Yes`"
        color="secondary"
        value="Yes"
        size="md"
        @click.native="vote = 'Yes'"
      />
      <TmBtn
        id="vote-no"
        :class="[vote === `No` ? 'active' : '']"
        :disabled="lastVoteOption === `No`"
        color="secondary"
        value="No"
        size="md"
        @click.native="vote = 'No'"
      />
      <TmBtn
        id="vote-veto"
        :class="[vote === `NoWithVeto` ? 'active' : '']"
        :disabled="lastVoteOption === `NoWithVeto`"
        color="secondary"
        value="No With Veto"
        size="md"
        @click.native="vote = 'NoWithVeto'"
      />
      <TmBtn
        id="vote-abstain"
        :class="[vote === `Abstain` ? 'active' : '']"
        :disabled="lastVoteOption === `Abstain`"
        color="secondary"
        value="Abstain"
        size="md"
        @click.native="vote = 'Abstain'"
      />
      <TmFormMsg
        v-if="$v.vote.$error && !$v.vote.required"
        name="Vote"
        type="required"
      />
    </TmFormGroup>
  </ActionModal>
</template>

<script>
import { mapGetters } from "vuex"
import { required } from "vuelidate/lib/validators"
import ActionModal from "./ActionModal"
import TmBtn from "src/components/common/TmBtn"
import TmFormGroup from "src/components/common/TmFormGroup"
import TmFormMsg from "src/components/common/TmFormMsg"

import transaction from "../utils/transactionTypes"

const isValid = option =>
  option === `Yes` ||
  option === `No` ||
  option === `NoWithVeto` ||
  option === `Abstain`

export default {
  name: `modal-vote`,
  components: {
    ActionModal,
    TmBtn,
    TmFormGroup,
    TmFormMsg
  },
  props: {
    proposalId: {
      type: [Number, String],
      required: true
    },
    proposalTitle: {
      type: String,
      required: true
    },
    lastVoteOption: {
      default: undefined,
      type: String
    }
  },
  data: () => ({
    vote: null
  }),
  computed: {
    ...mapGetters([`bondDenom`]),
    transactionData() {
      return {
        type: transaction.VOTE,
        proposalId: this.proposalId,
        option: this.vote
      }
    },
    notifyMessage() {
      return {
        title: `Successful vote!`,
        body: `You have successfully voted ${this.vote} on proposal #${this.proposalId}`
      }
    }
  },
  validations() {
    return {
      vote: {
        required,
        isValid
      }
    }
  },
  methods: {
    open() {
      this.$refs.actionModal.open()
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()

      this.vote = null
    }
  }
}
</script>

<style>
.proposal-title {
  font-weight: 500;
}

.action-modal-group.tm-form-group.vote-options {
  padding: 2rem 0;
}

.radio-container {
  display: flex;
  align-items: center;
  padding: 0.25rem 0;
}

.radio-container label {
  padding-left: 0.5rem;
}

.vote-options button {
  margin: 0;
  min-width: 49%;
}

.vote-options button span {
  margin: 0.25rem;
}

.vote-options button.active span {
  background: var(--tertiary);
}
</style>
