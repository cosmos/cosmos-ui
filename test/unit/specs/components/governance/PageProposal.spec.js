"use strict"

import moment from "moment"
import Vuelidate from "vuelidate"
import setup from "../../../helpers/vuex-setup"
import PageProposal from "renderer/components/governance/PageProposal"
import ModalDeposit from "renderer/components/governance/ModalDeposit"
import ModalVote from "renderer/components/governance/ModalVote"
import lcdClientMock from "renderer/connectors/lcdClientMock.js"

const {
  proposals,
  tallies,
  stakingParameters,
  governanceParameters
} = lcdClientMock.state
const proposal = proposals[`2`]

describe(`PageProposal`, () => {
  let wrapper, store
  const { mount, localVue } = setup()
  localVue.use(Vuelidate)
  localVue.directive(`tooltip`, () => {})
  localVue.directive(`focus`, () => {})

  const $store = {
    commit: jest.fn(),
    dispatch: jest.fn(),
    getters: {
      proposals: { proposals, tallies },
      bondDenom: stakingParameters.parameters.bond_denom,
      depositDenom: governanceParameters.deposit.min_deposit[0].denom
    }
  }

  beforeEach(() => {
    const instance = mount(PageProposal, {
      localVue,
      doBefore: ({ store }) => {
        store.commit(`setConnected`, true)
        store.commit(`setGovParameters`, governanceParameters)
        store.commit(`setStakingParameters`, stakingParameters.parameters)
        store.commit(`setProposal`, proposal)
        store.commit(`setProposalTally`, {
          proposal_id: `2`,
          tally_result: tallies[`2`]
        })
      },
      propsData: { proposalId: proposal.proposal_id },
      $store
    })
    wrapper = instance.wrapper
    store = instance.store
  })

  it(`has the expected html structure`, async () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`shows an error if the proposal couldn't be found`, () => {
    const instance = mount(PageProposal, {
      doBefore: ({}) => {
        store.commit(`setProposal`, {})
      },
      propsData: {
        proposalId: proposal.proposal_id
      },
      getters: {
        proposal: () => []
      }
    })

    wrapper = instance.wrapper
    store = instance.store
    store.commit(`setStakingParameters`, stakingParameters.parameters)

    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`should return the time of submission `, () => {
    expect(wrapper.vm.submittedAgo).toEqual(
      moment(new Date(proposal.submit_time)).fromNow()
    )
  })

  it(`should return the time that voting started`, () => {
    expect(wrapper.vm.votingStartedAgo).toEqual(
      moment(new Date(proposal.voting_start_block)).fromNow()
    )
  })

  it(`should return the time when deposits end`, () => {
    expect(wrapper.vm.depositEndsIn).toEqual(
      moment(new Date(proposal.deposit_end_time)).fromNow()
    )
  })

  describe(`Proposal status`, () => {
    it(`displays correctly a proposal that 'Passed'`, () => {
      wrapper.vm.proposal.proposal_status = `Passed`
      expect(wrapper.vm.status).toMatchObject({
        message: `This proposal has passed`
      })
    })

    it(`displays correctly a 'Rejected' proposal`, () => {
      wrapper.vm.proposal.proposal_status = `Rejected`
      expect(wrapper.vm.status).toMatchObject({
        message: `This proposal has been rejected and voting is closed`,
        color: `red`
      })
    })

    it(`displays correctly a proposal on 'DepositPeriod'`, () => {
      wrapper.vm.proposal.proposal_status = `DepositPeriod`
      expect(wrapper.vm.status).toMatchObject({
        message: `Deposits are open for this proposal`,
        color: `yellow`
      })
    })

    it(`displays correctly a proposal on 'VotingPeriod'`, () => {
      wrapper.vm.proposal.proposal_status = `VotingPeriod`
      expect(wrapper.vm.status).toMatchObject({
        message: `Voting for this proposal is open`,
        color: `green`
      })
    })

    it(`shows error status`, () => {
      wrapper.vm.proposal.proposal_status = ``
      expect(wrapper.vm.status).toMatchObject({
        message: `There was an error determining the status of this proposal.`,
        color: `grey`
      })
    })
  })

  describe(`Modal onVote`, () => {
    it(`enables voting if the proposal is on the 'VotingPeriod'`, () => {
      const proposal = proposals[`2`]
      proposal.proposal_status = `VotingPeriod`
      const instance = mount(PageProposal, {
        localVue,
        doBefore: ({ store }) => {
          store.commit(`setConnected`, true)
          store.commit(`setProposal`, proposal)
          store.commit(`setProposalTally`, {
            proposal_id: `2`,
            tally_result: tallies[`2`]
          })
        },
        propsData: {
          proposalId: proposal.proposal_id
        },
        $store
      })
      wrapper = instance.wrapper
      store = instance.store

      const voteBtn = wrapper.find(`#vote-btn`)
      voteBtn.trigger(`click`)

      expect(wrapper.vm.$store.dispatch.mock.calls).toEqual([
        [`getProposalVotes`, `2`]
      ])
      expect(wrapper.vm.lastVote).not.toBeDefined()
      expect(wrapper.contains(ModalVote)).toEqual(true)
      expect(voteBtn.html()).not.toContain(`disabled="disabled"`)
    })

    it(`load the last valid vote succesfully`, async () => {
      wrapper.setProps({ proposalId: `1` })
      wrapper.vm.wallet.address = lcdClientMock.state.votes[`1`][0].voter
      await wrapper.vm.onVote()
      expect(wrapper.vm.$store.dispatch.mock.calls).toEqual([
        [`getProposalVotes`, `1`]
      ])
      expect(wrapper.vm.lastVote).toBe(lcdClientMock.state.votes[`1`][0])
    })

    it(`keeps the last vote undefined if no vote to this proposal happened from the current address`, async () => {
      wrapper.setProps({ proposalId: `2` })
      await wrapper.vm.onVote()
      expect(wrapper.vm.$store.dispatch.mock.calls).toEqual([
        [`getProposalVotes`, `2`]
      ])
      expect(wrapper.vm.lastVote).toBe(undefined)
    })

    it(`disables voting if the proposal is on the 'DepositPeriod'`, () => {
      wrapper.setProps({ proposalId: `5` })
      expect(wrapper.find(`#vote-btn`).exists()).toEqual(false)
    })
  })

  describe(`Modal onDeposit`, () => {
    it(`enables deposits if the proposal on 'DepositPeriod'`, () => {
      const proposal = proposals[`5`]
      const instance = mount(PageProposal, {
        localVue,
        doBefore: ({ store }) => {
          store.commit(`setConnected`, true)
          store.commit(`setStakingParameters`, stakingParameters.parameters)
          store.commit(`setProposal`, proposal)
          store.commit(`setProposalTally`, {
            proposal_id: `5`,
            tally_result: tallies[`5`]
          })
        },
        propsData: {
          proposalId: proposal.proposal_id
        },
        $store
      })
      wrapper = instance.wrapper
      store = instance.store
      store.commit(`setGovParameters`, governanceParameters)
      wrapper.vm.proposal.proposal_status = `DepositPeriod`
      const depositBtn = wrapper.find(`#deposit-btn`)
      depositBtn.trigger(`click`)
      expect(wrapper.contains(ModalDeposit)).toEqual(true)
      expect(depositBtn.html()).not.toContain(`disabled="disabled"`)
    })

    it(`disables deposits if the proposal is not active`, () => {
      expect(wrapper.find(`#deposit-btn`).exists()).toEqual(false)
    })
  })

  it(`casts a vote`, async () => {
    wrapper.vm.$store.commit = jest.fn()
    wrapper.vm.$store.dispatch = jest.fn()

    await wrapper.vm.castVote({ option: `Abstain`, password: `12345` })

    expect(wrapper.vm.$store.dispatch.mock.calls).toEqual([
      [`submitVote`, { option: `Abstain`, proposal_id: `2`, password: `12345` }]
    ])

    expect(wrapper.vm.$store.commit.mock.calls).toEqual([
      [
        `notify`,
        {
          body: `You have successfully voted Abstain on proposal #2`,
          title: `Successful vote!`
        }
      ]
    ])
  })

  it(`shows an error if casting a vote fails`, async () => {
    wrapper.vm.$store.commit = jest.fn()
    wrapper.vm.$store.dispatch = jest.fn(() => {
      throw new Error(`unexpected error`)
    })

    await wrapper.vm.castVote({ option: `NoWithVeto`, password: `12345` })

    expect(wrapper.vm.$store.dispatch.mock.calls).toEqual([
      [
        `submitVote`,
        { option: `NoWithVeto`, proposal_id: `2`, password: `12345` }
      ]
    ])

    expect(wrapper.vm.$store.commit.mock.calls).toEqual([
      [
        `notifyError`,
        {
          body: `unexpected error`,
          title: `Error while voting on proposal #2`
        }
      ]
    ])
  })

  it(`allows the user to deposit on a proposal`, async () => {
    wrapper.vm.$store.commit = jest.fn()
    wrapper.vm.$store.dispatch = jest.fn()

    const amount = [
      {
        amount: `15`,
        denom: `atom`
      }
    ]

    await wrapper.vm.deposit({ amount, password: `12345` })

    expect(wrapper.vm.$store.dispatch.mock.calls).toEqual([
      [
        `submitDeposit`,
        {
          amount,
          proposal_id: `2`,
          password: `12345`
        }
      ]
    ])

    expect(wrapper.vm.$store.commit.mock.calls).toEqual([
      [
        `notify`,
        {
          body: `You have successfully deposited your STAKEs on proposal #2`,
          title: `Successful deposit!`
        }
      ]
    ])
  })

  it(`shows an error if depositing on a proposal fails`, async () => {
    wrapper.vm.$store.commit = jest.fn()
    wrapper.vm.$store.dispatch = jest.fn(() => {
      throw new Error(`unexpected error`)
    })

    const amount = [
      {
        amount: `9`,
        denom: `atom`
      }
    ]

    await wrapper.vm.deposit({ amount, password: `12345` })

    expect(wrapper.vm.$store.dispatch.mock.calls).toEqual([
      [
        `submitDeposit`,
        {
          amount,
          proposal_id: `2`,
          password: `12345`
        }
      ]
    ])

    expect(wrapper.vm.$store.commit.mock.calls).toEqual([
      [
        `notifyError`,
        {
          body: `unexpected error`,
          title: `Error while submitting a deposit on proposal #2`
        }
      ]
    ])
  })

  it(`disables interaction buttons if not connected`, () => {
    store.commit(`setConnected`, false)

    store.commit(
      `setProposal`,
      Object.assign({}, proposal, {
        proposal_status: `VotingPeriod`
      })
    )
    expect(
      wrapper.vm.$el.querySelector(`#vote-btn`).getAttribute(`disabled`)
    ).toBe(`disabled`)

    store.commit(
      `setProposal`,
      Object.assign({}, proposal, {
        proposal_status: `DepositPeriod`
      })
    )
    expect(
      wrapper.vm.$el.querySelector(`#deposit-btn`).getAttribute(`disabled`)
    ).toBe(`disabled`)
  })
})
