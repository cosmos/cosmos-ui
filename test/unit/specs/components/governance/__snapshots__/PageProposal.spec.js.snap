// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageProposal renders votes in HTML when voting is open 1`] = `
<tmpage-stub
  data-title="Proposal"
  subtitle=""
  title=""
>
  <div
    class="page-profile__header page-profile__section proposal"
  >
    <div
      class="row"
    >
      <h2
        class="proposal-id"
      >
        #2
      </h2>
       
      <div
        class="page-profile__header__info"
      >
        <div
          class="page-profile__status-and-title"
        >
          <span
            class="page-profile__status green"
          />
           
          <h2
            class="page-profile__title"
          >
            
              VotingPeriod proposal
            
          </h2>
        </div>
         
        <div
          class="page-profile__header__actions"
        >
          <tmbtn-stub
            color="primary"
            id="vote-btn"
            value="Vote"
          />
           
          <!---->
           
          <!---->
           
          <!---->
        </div>
      </div>
    </div>
     
    <div
      class="row"
    >
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Submitted
        </dt>
         
        <dd>
          a few seconds ago
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Proposal Status
        </dt>
         
        <dd>
          
            Voting started in 2 days
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Voting End Date
        </dt>
         
        <dd>
          January 5th 1970, 00:00
        </dd>
      </dl>
    </div>
  </div>
   
  <div
    class="page-profile__section"
  >
    <div
      class="row"
    >
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Deposit Count
        </dt>
         
        <dd>
          
            2,000
            STAKE
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Total Vote Count
        </dt>
         
        <dd>
          10,000
        </dd>
      </dl>
    </div>
     
    <div
      class="row"
    >
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Yes
        </dt>
         
        <dd>
          
            1,000 /
            10%
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          No
        </dt>
         
        <dd>
          
            2,000 /
            20%
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          No with Veto
        </dt>
         
        <dd>
          
            3,000 /
            30%
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Abstain
        </dt>
         
        <dd>
          
            4,000 /
            40%
          
        </dd>
      </dl>
    </div>
     
    <div
      class="row"
    >
      <div
        class="column"
      >
        <dl
          class="info_dl colored_dl"
        >
          <dt>
            Description
          </dt>
           
          <textblock-stub
            content="custom text proposal description"
          />
        </dl>
      </div>
    </div>
  </div>
   
  <modaldeposit-stub
    denom="STAKE"
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
   
  <modalvote-stub
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
</tmpage-stub>
`;

exports[`PageProposal should display proposal page if user has signed in 1`] = `
<tmpage-stub
  data-title="Proposal"
  subtitle=""
  title=""
>
  <div
    class="page-profile__header page-profile__section proposal"
  >
    <div
      class="row"
    >
      <h2
        class="proposal-id"
      >
        #2
      </h2>
       
      <div
        class="page-profile__header__info"
      >
        <div
          class="page-profile__status-and-title"
        >
          <span
            class="page-profile__status green"
          />
           
          <h2
            class="page-profile__title"
          >
            
              VotingPeriod proposal
            
          </h2>
        </div>
         
        <div
          class="page-profile__header__actions"
        >
          <tmbtn-stub
            color="primary"
            id="vote-btn"
            value="Vote"
          />
           
          <!---->
           
          <!---->
           
          <!---->
        </div>
      </div>
    </div>
     
    <div
      class="row"
    >
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Submitted
        </dt>
         
        <dd>
          a few seconds ago
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Proposal Status
        </dt>
         
        <dd>
          
            Voting started in 2 days
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Voting End Date
        </dt>
         
        <dd>
          January 5th 1970, 00:00
        </dd>
      </dl>
    </div>
  </div>
   
  <div
    class="page-profile__section"
  >
    <div
      class="row"
    >
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Deposit Count
        </dt>
         
        <dd>
          
            2,000
            STAKE
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Total Vote Count
        </dt>
         
        <dd>
          0
        </dd>
      </dl>
    </div>
     
    <div
      class="row"
    >
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Yes
        </dt>
         
        <dd>
          
            0 /
            0%
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          No
        </dt>
         
        <dd>
          
            0 /
            0%
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          No with Veto
        </dt>
         
        <dd>
          
            0 /
            0%
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Abstain
        </dt>
         
        <dd>
          
            0 /
            0%
          
        </dd>
      </dl>
    </div>
     
    <div
      class="row"
    >
      <div
        class="column"
      >
        <dl
          class="info_dl colored_dl"
        >
          <dt>
            Description
          </dt>
           
          <textblock-stub
            content="custom text proposal description"
          />
        </dl>
      </div>
    </div>
  </div>
   
  <modaldeposit-stub
    denom="STAKE"
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
   
  <modalvote-stub
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
</tmpage-stub>
`;

exports[`PageProposal should display proposal page should default tally to 0 if it's not yet present  1`] = `
<tmpage-stub
  data-title="Proposal"
  subtitle=""
  title=""
>
  <div
    class="page-profile__header page-profile__section proposal"
  >
    <div
      class="row"
    >
      <h2
        class="proposal-id"
      >
        #2
      </h2>
       
      <div
        class="page-profile__header__info"
      >
        <div
          class="page-profile__status-and-title"
        >
          <span
            class="page-profile__status green"
          />
           
          <h2
            class="page-profile__title"
          >
            
              VotingPeriod proposal
            
          </h2>
        </div>
         
        <div
          class="page-profile__header__actions"
        >
          <tmbtn-stub
            color="primary"
            id="vote-btn"
            value="Vote"
          />
           
          <!---->
           
          <!---->
           
          <!---->
        </div>
      </div>
    </div>
     
    <div
      class="row"
    >
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Submitted
        </dt>
         
        <dd>
          a few seconds ago
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Proposal Status
        </dt>
         
        <dd>
          
            Voting started in 2 days
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Voting End Date
        </dt>
         
        <dd>
          January 5th 1970, 00:00
        </dd>
      </dl>
    </div>
  </div>
   
  <div
    class="page-profile__section"
  >
    <div
      class="row"
    >
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Deposit Count
        </dt>
         
        <dd>
          
            2,000
            STAKE
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Total Vote Count
        </dt>
         
        <dd>
          0
        </dd>
      </dl>
    </div>
     
    <div
      class="row"
    >
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Yes
        </dt>
         
        <dd>
          
            0 /
            0%
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          No
        </dt>
         
        <dd>
          
            0 /
            0%
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          No with Veto
        </dt>
         
        <dd>
          
            0 /
            0%
          
        </dd>
      </dl>
       
      <dl
        class="info_dl colored_dl"
      >
        <dt>
          Abstain
        </dt>
         
        <dd>
          
            0 /
            0%
          
        </dd>
      </dl>
    </div>
     
    <div
      class="row"
    >
      <div
        class="column"
      >
        <dl
          class="info_dl colored_dl"
        >
          <dt>
            Description
          </dt>
           
          <textblock-stub
            content="custom text proposal description"
          />
        </dl>
      </div>
    </div>
  </div>
   
  <modaldeposit-stub
    denom="STAKE"
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
   
  <modalvote-stub
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
</tmpage-stub>
`;

exports[`PageProposal should display proposal page should show a loader if the necessary data hasen't been loaded 1`] = `
<tmpage-stub
  data-title="Proposal"
  subtitle=""
  title=""
>
  <tmdataloading-stub
    title="Data is loading…"
  />
</tmpage-stub>
`;

exports[`PageProposal shows an error if the proposal couldn't be found 1`] = `
<tmpage-stub
  data-title="Proposal"
  subtitle=""
  title=""
>
  <tmdataerror-stub />
</tmpage-stub>
`;
